grpc/creds/creds.go

```diff
@@ -0,0 +1,86 @@
+package creds
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/credentials"
+)
+
+// transportCredentials is a grpc/credentials.TransportCredentials which supports
+// connecting to, and verifying multiple DNS names
+type transportCredentials struct {
+	configs map[string]*tls.Config
+}
+
+// New returns a new initialized grpc/credentials.TransportCredentials
+func New(addrs []string, rootCAs *x509.CertPool, clientCerts []tls.Certificate) (credentials.TransportCredentials, error) {
+	configs := make(map[string]*tls.Config, len(addrs))
+	for _, addr := range addrs {
+		host, _, err := net.SplitHostPort(addr)
+		if err != nil {
+			return nil, err
+		}
+		configs[addr] = &tls.Config{
+			ServerName:   host,
+			RootCAs:      rootCAs,
+			Certificates: clientCerts,
+		}
+	}
+	return &transportCredentials{configs}, nil
+}
+
+// ClientHandshake performs the TLS handshake for a client -> server connection
+func (tc *transportCredentials) ClientHandshake(addr string, rawConn net.Conn, timeout time.Duration) (net.Conn, credentials.AuthInfo, error) {
+	config, present := tc.configs[addr]
+	if !present {
+		return nil, nil, fmt.Errorf("boulder/grpc/creds: Unexpected name, no TLS configuration present for \"%s\"", addr)
+	}
+	conn := tls.Client(rawConn, config)
+	errChan := make(chan error, 1)
+	go func() {
+		errChan <- conn.Handshake()
+	}()
+	var after <-chan time.Time
+	if timeout != 0 {
+		after = time.After(timeout)
+	}
+	select {
+	case <-after:
+		return nil, nil, errors.New("boulder/grpc/creds: TLS handshake timed out")
+	case err := <-errChan:
+		if err != nil {
+			_ = rawConn.Close()
+			return nil, nil, err
+		}
+		return conn, nil, nil
+	}
+}
+
+// ServerHandshake performs the TLS handshake for a server <- client connection
+func (tc *transportCredentials) ServerHandshake(rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	return nil, nil, fmt.Errorf("boulder/grpc/creds: Server-side handshakes are not implemented")
+}
+
+// Info returns information about the transport protocol used
+func (tc *transportCredentials) Info() credentials.ProtocolInfo {
+	return credentials.ProtocolInfo{
+		SecurityProtocol: "tls",
+		SecurityVersion:  "1.2",
```

Comment:
Should the `minVersion` and `maxVersion` be set in the `TLSConfig` instances created in `New` to match this hardcoded `SecurityVersion`? This might be confusing in a world where TLS 1.3 support is added and our GRPC connections use it, but `Info` claims otherwise.